stellargraph/datasets/dataset_loader.py

```diff
@@ -0,0 +1,108 @@
+        url: str,
+        url_archive_format: Optional[str],
+        expected_files: List[str],
+        description: str,
+        source: str,
+        data_subdirectory_name: Optional[str] = None,
+    ) -> None:
+        self.name = name
+        self.directory_name = directory_name
+        self.url = url
+        self.url_archive_format = url_archive_format
+        self.expected_files = expected_files
+        self.description = description
+        self.source = source
+        self.data_subdirectory_name = data_subdirectory_name
+
+    @property
+    def base_directory(self) -> str:
+        """Return the path of the data directory for this dataset"""
+        return os.path.join(self._all_datasets_directory(), self.directory_name)
+
+    @property
+    def data_directory(self) -> str:
+        """Return the directory containing the data content files"""
+        if self.data_subdirectory_name is None:
+            return self.base_directory
+        else:
+            return os.path.join(self.base_directory, self.data_subdirectory_name)
+
+    def _create_base_directory(self) -> None:
+        data_dir = self.base_directory
+        if not os.path.exists(data_dir):
+            os.makedirs(data_dir)
+
+    @staticmethod
+    def _all_datasets_directory() -> str:
+        """Return the path of the base directory which contains subdirectories for each dataset."""
+        return os.path.expanduser(
+            os.getenv("STELLARGRAPH_DATASETS_PATH", os.path.join("~", "data"))
+        )
+
+    def _is_downloaded(self) -> bool:
+        """Returns true if the expected files for the dataset are present"""
+        for file in self.expected_files:
+            if not os.path.isfile(os.path.join(self.base_directory, file)):
+                return False
+        return True
+
+    def download(self, ignore_cache: Optional[bool] = False) -> None:
+        """Download the dataset (if not already downloaded, unless ignore_cache=True)"""
+        if not self._is_downloaded() or ignore_cache:
+            print(
+                f"{self.name} dataset downloading to {self.base_directory} from {self.url}"
+            )
+            if self.url_archive_format is None:
+                # single file to download
+                self._create_base_directory()
+                destination_filename = os.path.join(
+                    self.base_directory, self.expected_files[0]
+                )
+                urlretrieve(self.url, filename=destination_filename)
+            else:
+                # archive of files
+                filename, _ = urlretrieve(
+                    self.url
+                )  # this will download to a temporary location
+                self._create_base_directory()
+                unpack_archive(
+                    filename, self._all_datasets_directory(), self.url_archive_format
+                )
+            if not self._is_downloaded():
+                print(f"{self.name} dataset failed to download")
```

Comment:
I think this should be an error:

```suggestion
                raise ...(f"{self.name} dataset failed to download")
```

It might be good to be more specific in the error message too, e.g. try to determine why we think it failed. I don't know what exception would be most appropriate, but maybe `FileNotFoundError` if this only ever fails because of missing files.

Reply:
Now using FileNotFoundError with a message listing the missing files